cpp/src/stream_compaction/apply_boolean_mask.cu

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "copy_if.cuh"
+#include <cudf/table.hpp>
+ 
+namespace {
+
+// Returns true if the mask is true and valid (non-null) for index i
+// This is the filter functor for apply_boolean_mask
+// Note we use a functor here so we can cast to a bitmask_t __restrict__
+// pointer on the host side, which we can't do with a lambda.
+template <bool has_data, bool has_nulls>
+struct boolean_mask_filter
+{
+  boolean_mask_filter(gdf_column const & boolean_mask) :
+    size{boolean_mask.size},
+    data{reinterpret_cast<cudf::bool8 *>(boolean_mask.data)},
+    bitmask{reinterpret_cast<bit_mask_t *>(boolean_mask.valid)}
+    {}
+
+  __device__ inline 
+  bool operator()(gdf_index_type i)
+  {
+    if (i < size) {
```

Comment:
I was thinking that for arbitrary filter functors the kernel can't decide whether or not to run the filter... However I guess it can at least *not* run it on indices larger than the column size.

Reply:
Actually moving the bounds check outside the operator doesn't save anything. I have to call the operator like this: `bool mask_true = (tid < size) && filter(tid);` in all cases, because the kernels that call it have `__syncthreads()` so my grid-stride loops can't early out.  But it simplifies the operators a bit.